include/zephyr/spinlock.h

```diff
@@ -199,16 +227,34 @@ static ALWAYS_INLINE int k_spin_trylock(struct k_spinlock *l, k_spinlock_key_t *
 
 	z_spinlock_validate_pre(l);
 #ifdef CONFIG_SMP
+#ifdef CONFIG_TICKET_SPINLOCKS
+	atomic_val_t ticket_val = atomic_get(&l->owner);
+
+	/*
+	 * For real-life systems the spinlock values will not be able to wrap
+	 * around to erroneously match the ticket_val, so if the CAS succeeds
+	 * we may be sure that the spinlock wasn't locked.
+	 */
+	if (!atomic_cas(&l->tail, ticket_val, ticket_val + 1)) {
+		goto busy;
+	}
```

Comment:
The code looks fine, but the comment is really misleading. I'm sure that `For real-life systems the spinlock values will not be able to wrap around` statement isn't correct.

However such wrap-around should work fine even with existing code. So, let's test it with additional test-case in atomics test as was proposed by @npitre:

```
the best way to put those worries to rest is to write a test for that particular behavior. If some architecture doesn't wrap
integers by default, then it is likely that it won't wrap atomic increments either. So you are not necessarily more theoretically "correct" by doing the increment atomically. Again, adding a test for that as well is the best safeguard against unexpected deviation from widespread practice.
```

Reply:
Hi @evgeniy-paltsev , I think there is a kind of confusion here, looks like we mean 2 different things:
1) Under "wrap around" here I meant that someone will take and release the spinlock 0xffffffff+  times after we read the ticket_val and before we do atomic_cas. So that the spinlock value will actually match our expectation of  a free spinlock while being held by someone in reality. @andyross suggested a malicious attacker who could try this.
My impression is that we can defend ourselves from such kind of attack at the cost of extra memory (e.g. one more atomic variable), but this probably may come as a following up feature if such scenario will be considered really critical.
2) As for the integer wrap-around, I agree, let's have a separate test on this to put our mind at rest

Reply:
Hi @amr-sc ,

> Under "wrap around" here I meant that someone will take and release the spinlock 0xffffffff+ times after we read the ticket_val and before we do atomic_cas. So that the spinlock value will actually match our expectation of a free spinlock while being held by someone in reality. @andyross suggested a malicious attacker who could try this.
My impression is that we can defend ourselves from such kind of attack at the cost of extra memory (e.g. one more atomic variable), but this probably may come as a following up feature if such scenario will be considered really critical.

Yep, now I understand that I was talking about different kind of overflow. And it's indeed the reason we should change the comment to make it less confusing :)